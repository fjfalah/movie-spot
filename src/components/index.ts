export { default as Box } from './Box';
export { default as Button } from './Button';
export { default as GlobalStyles } from './GlobalStyles';
export { default as InfiniteScroll } from './InfiniteScroll';
export { default as Input } from './Input';
export { default as InputSearch } from './InputSearch';
export { default as Loading } from './Loading';
export { default as MovieCard } from './MovieCard';
export { default as PageDescription } from './PageDescription';
export { default as Pill } from './Pill';
export { default as Section } from './Section';
export { default as Text } from './Text';
