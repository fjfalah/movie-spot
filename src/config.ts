export const IS_PRODUCTION = process.env.NODE_ENV === 'production';
export const API_URL = process.env.API_URL || 'http://www.omdbapi.com';
export const API_URL_KEY = process.env.API_URL_KEY || 'faf7e5bb';
