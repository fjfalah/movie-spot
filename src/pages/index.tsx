import { NextPage } from 'next';
import React, { useEffect, useCallback } from 'react';
import { connect, useDispatch } from 'react-redux';

import {
  Section,
  PageDescription,
  MovieCard,
  InfiniteScroll,
  Loading,
  InputSearch,
  Text,
  Box,
} from '../components';
import { getMovies } from '../services/movieService';
import { searchMoviesAction } from '../store/actions/movieAction';
import styled, { theme } from '../themes';
import { MovieStateType } from '../types/state';

type HomeType = {
  movies: MovieStateType;
};

const HeaderWrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;

  @media only screen and (max-width: 720px) {
    flex-direction: column;
  }
`;

const MovieWrapper = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
`;

const MovieNotFound = styled.div`
  display: flex;
  flex-direction: row;
  padding: 20px;
`;

const HomePage: NextPage<HomeType> = ({ movies }) => {
  const dispatch = useDispatch();
  const { keyword, items, page, hasMore } = movies;

  const handleLoadMore = useCallback(() => {
    dispatch(getMovies(keyword, page + 1));
  }, [dispatch, keyword, page]);

  const handleSubmitSearch = useCallback(
    (e) => {
      e.preventDefault();
      const data = new FormData(e.target);
      const search = data.get('search').toString();
      dispatch(searchMoviesAction(search));
    },
    [dispatch]
  );
  useEffect(() => {
    if (items.length === 0) {
      dispatch(getMovies(keyword, 1));
    }
  }, [dispatch, keyword, items]);
  return (
    <PageDescription title="Home">
      <Section>
        <HeaderWrapper>
          <Text size="30px" color={theme.color.white} weight="bold">
            MovieSpot
          </Text>
          <Box height="30px" />
          <InputSearch
            onSubmit={handleSubmitSearch}
            placeholder="Search by Title"
            defaultValue={keyword}
          />
        </HeaderWrapper>
        <Box height="30px" />
        {items.length === 0 && (
          <MovieNotFound>
            <Text size="20px" color={theme.color.white} weight="bold">
              No Movie Found
            </Text>
          </MovieNotFound>
        )}
        <InfiniteScroll
          onLoadMore={handleLoadMore}
          hasMore={hasMore}
          loader={<Loading />}
        >
          <MovieWrapper>
            {(movies?.items || []).map((item) => {
              const { Poster, Title, Type, Year, imdbID } = item;
              return (
                <MovieCard
                  title={Title}
                  imdbID={imdbID}
                  poster={Poster}
                  year={Year}
                  type={Type}
                  key={imdbID}
                />
              );
            })}
          </MovieWrapper>
        </InfiniteScroll>
      </Section>
    </PageDescription>
  );
};

const mapStateToProps = (state) => ({
  movies: state.movies,
});

export default connect(mapStateToProps)(HomePage);
