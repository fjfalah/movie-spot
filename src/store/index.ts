import { useMemo } from 'react';
import {
  createStore,
  applyMiddleware,
  Store,
  $CombinedState,
  AnyAction,
} from 'redux';
import logger from 'redux-logger';
import thunk from 'redux-thunk';

import { IS_PRODUCTION } from '../config';

import reducers from './reducers';

type ConfigureStoreType = Store<
  {
    readonly [$CombinedState]?: undefined;
  },
  AnyAction
> & {
  dispatch: unknown;
};

let store;
const middlewares = [];
if (!IS_PRODUCTION) {
  middlewares.push(logger);
}
middlewares.push(thunk);
const configureStore = (initialState = {}): ConfigureStoreType => {
  return createStore(reducers, initialState, applyMiddleware(...middlewares));
};
export const initializeStore = (preloadedState = {}): ConfigureStoreType => {
  let newStore = store ?? configureStore(preloadedState);

  // After navigating to a page with an initial Redux state, merge that state
  // with the current state in the store, and create a new store
  if (preloadedState && store) {
    newStore = configureStore({
      ...store.getState(),
      ...preloadedState,
    });
    // Reset the current store
    store = undefined;
  }

  // For SSG and SSR always create a new store
  if (typeof window === 'undefined') return newStore;
  // Create the store once in the client
  if (!store) store = newStore;

  return newStore;
};

export const useStore = (initialState = {}): ConfigureStoreType => {
  return useMemo(() => initializeStore(initialState), [initialState]);
};

export type AppDispatch = typeof store.dispatch;
