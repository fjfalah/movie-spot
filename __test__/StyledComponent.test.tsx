import { mount } from 'enzyme';
import React from 'react';
import { ThemeProvider } from 'styled-components';
import 'jest-styled-components';

import {
  Button,
  Box,
  Input,
  InputSearch,
  Pill,
  Section,
  Text,
  MovieCard,
} from '../src/components';
import { theme } from '../src/themes';

const mountWithTheme = (component: React.ReactNode) => {
  return mount(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
};

describe('Test - Component', () => {
  it('Theme Provider should works with checking color', () => {
    const wrapper = mountWithTheme(<Button>test</Button>);
    expect(wrapper.debug()).toMatchSnapshot();
    expect(wrapper).toHaveStyleRule('background', '#1b998b');
  });
  it('Box Should has width and height', () => {
    const wrapper = mountWithTheme(
      <Box width="10px" height="10px" backgroundColor="red" />
    );

    expect(wrapper.children().debug()).toMatchSnapshot();
    expect(wrapper).toHaveStyleRule('background-color', 'red');
    expect(wrapper).toHaveStyleRule('width', '10px');
    expect(wrapper).toHaveStyleRule('height', '10px');
  });
  it('Pill Should has color and background value', () => {
    const wrapper = mountWithTheme(
      <Pill color="#ffffff" background="red">
        Batman
      </Pill>
    );

    expect(wrapper.children().debug()).toMatchSnapshot();
    expect(wrapper).toHaveStyleRule('background', 'red');
    expect(wrapper).toHaveStyleRule('color', '#ffffff');
  });
  it('Text Should has color, size and weight value', () => {
    const wrapper = mountWithTheme(
      <Text color="#000000" size="20px" weight="bold">
        Hello, World!
      </Text>
    );

    expect(wrapper.children().debug()).toMatchSnapshot();
    expect(wrapper).toHaveStyleRule('color', '#000000');
    expect(wrapper).toHaveStyleRule('font-size', '20px');
    expect(wrapper).toHaveStyleRule('font-weight', 'bold');
  });
  it('Input should has value', () => {
    const wrapper = mountWithTheme(<Input defaultValue="Batman" />);

    expect(wrapper.find('input').debug()).toMatchSnapshot();
    expect(wrapper.find('input').props().defaultValue).toEqual('Batman');
  });
  it('InputSearch calls onSubmit prop function when form is submitted', () => {
    const onSubmitFn = jest.fn();
    const wrapper = mountWithTheme(
      <InputSearch onSubmit={onSubmitFn} defaultValue="Batman" />
    );
    const form = wrapper.find('form');
    form.simulate('submit');
    expect(onSubmitFn).toHaveBeenCalledTimes(1);
    expect(form.debug()).toMatchSnapshot();
  });
  it('Section should has children', () => {
    const wrapper = mountWithTheme(
      <Section>
        <Box />
        <Box />
      </Section>
    );
    expect(wrapper.find('section').debug()).toMatchSnapshot();
    expect(wrapper.find('section').children()).toHaveLength(2);
  });
  it('MovieCard should has props', () => {
    const props = {
      poster:
        'https://m.media-amazon.com/images/M/MV5BYThjYzcyYzItNTVjNy00NDk0LTgwMWQtYjMwNmNlNWJhMzMyXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg',
      imdbID: '123',
      year: `2016`,
      type: 'test',
      title: 'Batman vs Superman',
    };
    const wrapper = mountWithTheme(
      <MovieCard
        imdbID={props.imdbID}
        poster={props.poster}
        title={props.title}
        year={props.year}
        type={props.type}
      />
    );
    expect(wrapper.children().debug()).toMatchSnapshot();
    expect(wrapper.find('.title').first().text()).toEqual(props.title);
    expect(wrapper.find('.poster').first()).toHaveStyleRule(
      'background-image',
      `url(${props.poster})`
    );
    expect(wrapper.find('label').text()).toEqual(props.year);
    expect(wrapper.find('a').at(0).props().href).toEqual(`/${props.imdbID}`);
  });
});
