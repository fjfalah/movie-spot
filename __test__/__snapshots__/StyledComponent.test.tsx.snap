// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test - Component Box Should has width and height 1`] = `
"<styled.div width=\\"10px\\" height=\\"10px\\" backgroundColor=\\"red\\">
  <div width=\\"10px\\" height=\\"10px\\" className=\\"sc-AxjAm kLHCDb\\" />
</styled.div>"
`;

exports[`Test - Component Input should has value 1`] = `"<input defaultValue=\\"Batman\\" className=\\"sc-AxhCb kyYauG\\" />"`;

exports[`Test - Component InputSearch calls onSubmit prop function when form is submitted 1`] = `
"<form onSubmit={[Function: mockConstructor]} className=\\"sc-AxhUy iRkoav\\">
  <Styled(styled.input) type=\\"text\\" name=\\"search\\" defaultValue=\\"Batman\\">
    <input type=\\"text\\" name=\\"search\\" defaultValue=\\"Batman\\" className=\\"sc-AxhCb sc-AxheI gVBdlz\\" />
  </Styled(styled.input)>
  <Styled(styled.button) type=\\"submit\\">
    <button type=\\"submit\\" className=\\"sc-AxirZ sc-AxgMl bYfJPD\\">
      Search
    </button>
  </Styled(styled.button)>
</form>"
`;

exports[`Test - Component MovieCard should has props 1`] = `
"<MovieCard imdbID=\\"123\\" poster=\\"https://m.media-amazon.com/images/M/MV5BYThjYzcyYzItNTVjNy00NDk0LTgwMWQtYjMwNmNlNWJhMzMyXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg\\" title=\\"Batman vs Superman\\" year=\\"2016\\" type=\\"test\\">
  <styled.div>
    <div className=\\"sc-fzoLsD fhjtzk\\">
      <styled.div url=\\"https://m.media-amazon.com/images/M/MV5BYThjYzcyYzItNTVjNy00NDk0LTgwMWQtYjMwNmNlNWJhMzMyXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg\\" onClick={[undefined]} className=\\"poster\\">
        <div onClick={[undefined]} className=\\"sc-fzpans hwcKCS poster\\" />
      </styled.div>
      <styled.label>
        <label className=\\"sc-fzplWN ZLyjU\\">
          2016
        </label>
      </styled.label>
      <styled.div>
        <div className=\\"sc-fznyAO gEnKLi\\">
          <styled.div>
            <div className=\\"sc-fznKkj fQiBHL\\">
              <Styled(styled.p) size=\\"16px\\" weight=\\"bold\\" className=\\"title\\">
                <p size=\\"16px\\" className=\\"sc-fzozJi sc-fznZeY czkRfR title\\">
                  Batman vs Superman
                </p>
              </Styled(styled.p)>
            </div>
          </styled.div>
          <styled.div>
            <div className=\\"sc-fzokOt eXkiqi\\">
              <Link href=\\"/[id]\\" as=\\"/123\\" passHref={true}>
                <a onClick={[Function: onClick]} onMouseEnter={[Function (anonymous)]} href=\\"/123\\">
                  <styled.button>
                    <button className=\\"sc-AxirZ bdZEKS\\">
                      DETAIL
                    </button>
                  </styled.button>
                </a>
              </Link>
            </div>
          </styled.div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</MovieCard>"
`;

exports[`Test - Component Pill Should has color and background value 1`] = `
"<styled.label color=\\"#ffffff\\" background=\\"red\\">
  <label color=\\"#ffffff\\" className=\\"sc-fzqBZW iJKlZl\\">
    Batman
  </label>
</styled.label>"
`;

exports[`Test - Component Section should has children 1`] = `
"<section className=\\"sc-fzqNJr fgapPO\\">
  <styled.div>
    <div className=\\"sc-AxjAm gINqhJ\\" />
  </styled.div>
  <styled.div>
    <div className=\\"sc-AxjAm gINqhJ\\" />
  </styled.div>
</section>"
`;

exports[`Test - Component Text Should has color, size and weight value 1`] = `
"<styled.p color=\\"#000000\\" size=\\"20px\\" weight=\\"bold\\">
  <p color=\\"#000000\\" size=\\"20px\\" className=\\"sc-fzozJi iMRFzr\\">
    Hello, World!
  </p>
</styled.p>"
`;

exports[`Test - Component Theme Provider should works with checking color 1`] = `
"<ThemeProvider theme={{...}}>
  <styled.button>
    <button className=\\"sc-AxirZ bdZEKS\\">
      test
    </button>
  </styled.button>
</ThemeProvider>"
`;
